### ISSUE TITLE
allowedNetworks doesn't work with IPv6 addresses

### ISSUE THREAD

**Issue by trapperhoney** [syncthing/syncthing#7821]
### Does your log mention database corruption?

No

### Include required information

Please be sure to include at least:

 - which version of Syncthing and what operating system you are using
1.12.1~ds1-2 from Ubuntu 21.04
 - browser and version, if applicable
 N/A
 - what happened,
Use of allowedNetworks always prevents connectivity in an IPv6-only environment.
 - what you expected to happen instead, and
 allowedNetworks should work as described in https://docs.syncthing.net/advanced/device-allowednetworks.html and as it does on an IPv4 capable network.
 - any steps to reproduce the problem.
 
I managed to boil this down to a fairly simple example, disabling all of the dynamic discovery features and using explicit IP addresses instead of DNS names (though, the issue was happening with those enabled as well).

First, setup two devices which try to connect to each other directly using the 'address' configuration field.  The address is of the form: tcp6://[2001:peers address].  Verify connectivity.

Now, change one of the peers to use a different, non existent address.  Restart both servers.  The connection will still establish since one peer can still initiate.  This step is to make sure we know who the initiator is and better demonstrate what I think the underlying problem is...

Now, on the good initiator, add an allowedNetworks argument.  Even ::/0 is sufficient to demonstrate this problem.  Restart that server.  Now, it won't even try to initiate the connection.  The UI shows the error 'network disallowed'.  The network debug log shows "Network for tcp6://[2001:peers ip] is disallowed".

This leads me to believe logic for validating that the destination IP address in the context of the allowedNetworks list within the initiator is broken.  And indeed, putting an allowedNetworks block in the receiver without one in the initiator works just fine.

Doing the same thing with IPv4 addresses works as expected in both directions.

**Comment by AudriusButkevicius** [comment]
I think Ipv6 addresses encoded in urls need to be surrounded by square braces. Did you do that? I think this is best moved to the forum until we confirm that it's an actual bug rather than holding it wrong kind of thing.

**Comment by ProactiveServices** [comment]
You need to drop the "6" in your device address spec. I also get connection failures[0] when I use "tcp6://[IPv6 address]". I wonder if the UI should prevent this sort of input? There's no need for you to disable global discovery and use specific addresses - even with global discovery enabled, allowedNetworks takes precedence.

0: received unexpected handshake message of type *tls.clientHelloMsg when waiting for *tls.serverHelloMsg (23:54:41)

Also you're six months behind on updates, 1.12 is pretty old now.

**Comment by trapperhoney** [comment]
Thanks for the pointers.  I've tried both removing the the '6' from tcp and using [] for the network prefix (though, the documentation explicitly shows non-bracketed IPv6 subnets).  Neither of these suggestions helped, though :\

**Comment by calmh** [comment]
This works as advertised for me. Setting a single allowed entry as `fe80::/16` prevents all IPv4 connections with "network disallowed" but allows the link-local v6 to happen..

<img width="570" alt="Screen Shot 2021-07-16 at 06 30 34" src="https://user-images.githubusercontent.com/125426/125892074-efc866f6-c127-4ab8-929c-a7dc37126f3a.png">

**Comment by calmh** [comment]
I suggest enabling `connections` and `dialer` tracing and seeing what the log shows, posting the resulting things if it indicates filtering applied badly.

**Comment by trapperhoney** [comment]
> This works as advertised for me. Setting a single allowed entry as `fe80::/16` prevents all IPv4 connections with "network disallowed" but allows the link-local v6 to happen..

Does it work when you use this setting on _both_ sides?  That's the real bug here, sorry if that wasn't clear from above... Those instructions intentionally cripple one side's ability to initiate the connection so that you know which side is starting the connection.  Then, manipulating the allowedNetworks field on the side that can still establish a connection shows that the bug is in the code that (from this [doc](https://docs.syncthing.net/advanced/device-allowednetworks.html)):
- Attempt connections to addresses in the specified networks (manually configured or discovered).
- Not attempt connections to addresses outside the specified networks, regardless of whether manually configured or automatically discovered.

I previously tried enabled verbose logging before, but there's not much useful there:
```
Jul 16 09:12:20 DEBUG: Reconnect loop
Jul 16 09:12:20 DEBUG: Reconnect loop for X4AZTA5-... [tcp6://[2001:...]]
Jul 16 09:12:20 DEBUG: Not dialing X4AZTA5-... via tcp6://[2001:...] as sleep is 1s, next dial is at 2021-07-16 09:13:19.200108238 -0400 EDT m=+61.463745128 and current time is 2021-07-16 09:12:20.202647508 -0400 EDT m=+2.466284401
Jul 16 09:12:20 DEBUG: initial rampup; sleep 2s and update to 4s
```

**Comment by AudriusButkevicius** [comment]
The setting works both ways. 

It will only dial addresses matching the allowed list, and will only accept incoming connections matching the allowed list.

**Comment by trapperhoney** [comment]
I dug into the code a bit and I think I found the problem.  The issues all boil down to the use of `net.ResolveIPAddr` in `IsAllowedNetwork` [here](https://github.com/syncthing/syncthing/blob/ae30b46bfef287a029702bf7c494fd7bab46dc2c/lib/connections/service.go#L972)

First, this function can only parse an IPv6 address if either:
- The address _does not_ include square brackets nor a port.  eg 2001::1
- The address includes square brackets _and_ a port.  eg [2001::1]:22000

The address configuration parameter only allows for IPv6 addresses with square bracket notation.  It does not require a port (and defaults to 22000), but the `net.ResolveIPAddr` function requires a port when using square brackets.  Adding port 22000 to my initial example fixes the problem.

However, there's a more fundamental issue here.  Most people will use a host name in the address field, not the IP address explicitly.  And indeed the [documentation](https://docs.syncthing.net/users/config.html) for the address field explicitly supports this.  The [documentation](https://pkg.go.dev/net#ResolveIPAddr) for the `net.ResolveIPAddr` function warns against this, though:

> The address parameter can use a host name, but this is not recommended, because
> it will return at most one of the host name's IP addresses.

In my case, when using a host name, multiple IP addresses are returned in the resolution and the `net.ResolveIPAddr` function happens to pick the IPv4 address (which fails the 'allow' check) instead of the IPv6 address.  I imagine this would be an issue if multiple IPv4 or multiple IPv6 addresses were returned as well (but the dual-stack case seems like it would be the more common problem).

So, to summarize, you can work around this issue by explicitly setting an IPv6 address and port in the address field, but the allowedNetworks feature is fundamentally broken when used with a host name that can return multiple IP addresses.  There doesn't seem to be an obvious fix for this using only the go net package...

**Comment by AudriusButkevicius** [comment]
I guess my vote would be to change the documentation to explain this caveat and encourage people to use ip addresses.

**Comment by AudriusButkevicius** [comment]
Actually, https://pkg.go.dev/net#LookupIP should work.

**Comment by trapperhoney** [comment]
Indeed, sorry I missed that one... The only tricky part would be ensuring that the address used during this check is actually the same one used when establishing the connection.  Maybe a slightly more robust fix would be to pass in the resolved address for the allow check?  I haven't finished parsing how the connection establishment code does this yet, though...

**Comment by AudriusButkevicius** [comment]
But this would still not do what you want, because we would dial only one of the resolved addresses, not call, so you can still hit a case where you resolve a v4 address rather than a v6 one which does not match the allowed networks.

**Comment by trapperhoney** [comment]
Right, the LookupIP call and filtering based on allowedNetwork would need to be done first, and then allow Dial to proceed with whatever address passed that filter (if any).  This would be a more invasive change than simply modifying the behavior of IsAllowedNetwork...

**Comment by calmh** [comment]
I'm not following the part about brackets and ports. You can specify addresses in a few different ways, yes, but this has nothing to do with allowedNetworks which is a list of ... networks. We don't do string comparison between the two. 

As for host names I guess we should dial all resolved addresses (modulo allowedNetworks etc of course).

**Comment by trapperhoney** [comment]
> I'm not following the part about brackets and ports. You can specify addresses in a few different ways, yes, but this has nothing to do with allowedNetworks which is a list of ... networks. We don't do string comparison between the two.

Sorry, I may have written this part too quickly... The `IsAllowedNetwork` function calls `net.SplitHostPort(host)` on the address.  When the address is of the form [2001:...]:22000, `host` is updated to "2001:..." (without the port nor the brackets) before the resolution is done.  When no port is in the address (eg [2001:...]), the split function fails and and the host string remains [2001:...] with the brackets.  This bracketed version causes `net.ResolveIPAddr` to fail.

So since the address configuration field requires brackets but not a port, and IsAllowedNetwork expects a port, not using a port breaks the attempt at resolution.

**Comment by trapperhoney** [comment]
Perhaps a better way of illustrating the bracketing issue is the following additional test case in TestAllowedNetworks:
```
diff --git a/lib/connections/connections_test.go b/lib/connections/connections_test.go
index b9a10d5caa50..ca753cfc1d01 100644
--- a/lib/connections/connections_test.go
+++ b/lib/connections/connections_test.go
@@ -93,6 +93,11 @@ func TestAllowedNetworks(t *testing.T) {
                        []string{"192.168.0.0/24", "fe80::/48"},
                        true,
                },
+               {
+                       "[fe80::1]",
+                       []string{"192.168.0.0/24", "fe80::/48"},
+                       true,
+               },
                {
                        "192.168.0.1",
                        []string{"192.168.1.0/24", "fe80::/48"},
```

The test case that comes just before this new one uses a host of "fe80::1", but as far as I can tell, there's no way to actually specify an IPv6 IP address in the address configuration option without using brackets...

**Comment by AudriusButkevicius** [comment]
I think the idea is that brackes are not used for CIDRs, but are/must be used for addresses of devices, but I think the parsing already handles this?

**Comment by calmh** [comment]
OK, this is in fact a legit bug where something isn't resolved properly at some stage.

<img width="875" alt="Screenshot 2021-07-16 at 21 14 31" src="https://user-images.githubusercontent.com/125426/125997761-d3b49e23-50fe-4c87-b2a3-51cc1ef8237a.png">


```
2021-07-16 21:12:33 Resolved device PSEUDOP-YLA5DBR-XO2HBNT-2FOG4AC-SWOW5EZ-H3QQY76-GISB7PG-E2SWOA2 addresses: [quic://172.16.32.11:22000 quic://83.233.124.224:22000 tcp://172.16.32.11:22000 tcp://83.233.124.224:13611 tcp://83.233.124.224:22000 tcp://83.233.124.224:62069 tcp://[2001:db8::1]]
[...]
2021-07-16 21:12:33 Network for tcp://[2001:db8::1] is disallowed
```
